### DISCUSSION TITLE
Is there a good practice for using Signal?

### DISCUSSION THREAD

**Discussion by ZephyrZhang3** [DioxusLabs/dioxus#2263] [Q&A]
I'm new about Dioxus for two days and started directly from 0.5. 

When I read the document(just the guide and a piece of reference), the part about signal was really confusing. 
`signal`, `signal =`, `signal()`, `*signal.read()`, `signal.read().clone()`, `signal.write()`, `*signal.write() =`, `signal().write().push()`, `signal.push()`, ..., there are too many ways to use it. 

The documentation doesn’t make it very clear, and the code examples also use them mixedly. 
I think the key is about signal is `Copy` or not, but still confused in using.

So, is there some good practice for using signal?

I watching into the source and docs.rs, and find that
[ `signal()` is just like the `signal.read().clone()`](https://docs.rs/dioxus-signals/0.5.1/src/dioxus_signals/read.rs.html#13-151:~:text=%23%5Bdoc(hidden)%5D,%26_%0A%20%20%20%20%7D%0A%7D), but does this two way the same? Eveytime I call signal(), it cloned?

[ the `signal.set(value)` is calling `*signal.write() = value`](https://docs.rs/dioxus-signals/0.5.1/src/dioxus_signals/write.rs.html#68-73:~:text=%23%5Btrack_caller%5D%0A%20%20%20%20fn%20set(%26mut%20self%2C%20value%3A%20Self%3A%3ATarget)%0A%20%20%20%20where%0A%20%20%20%20%20%20%20%20Self%3A%3ATarget%3A%20Sized%2C%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20*self.write()%20%3D%20value%3B%0A%20%20%20%20%7D). Actually, the `set()` is in doc's `Migration` and `dynamic rendering` chapter, but I find it in docs.rs firstly🤣. It seems `set()` is exist in before verson. I think this should be in the `Guide` and clearly explained.

And there are `Vec` and `Option` signal methods in docs.rs doc. BUT this seems not in doc(maybe I don't find?)

Anyway, the doc is not frendly enough for me😢.

**Comment by frantisek-heca** [comment]
yeah, had the same struggles few days ago when comming new to dioxus and reading the actual docs.
unfortunately i resolved most of my questions at discord, which i don't like to promote in general, because these web discussion are the best for searchability. 
i have my perosnal chaotic docs (screenshots from the discord), that i am not sure if to share here, but because we have nothing better, it could be helpful.

for a start.. this is my conclusion of alternative ways to use signals
```rust
count += 1;
*count.write() = 1;
*count.write() += 1;  
*count.write() = *count.read() + 1;

*count.write() = count() + 1;
*count.write() = count.cloned() + 1;
*count.write() = count.read().clone() + 1;

// every signal has .set(value) method, that calls `*self.write() = value` inside
count.set(1);
```

and this could be helpful...
<img width="1222" alt="image" src="https://github.com/DioxusLabs/dioxus/assets/25659432/66014c7d-b75f-4c04-bc4c-ecb74486ba54">

`count += 1` works because of this internal implementation
<img width="420" alt="image" src="https://github.com/DioxusLabs/dioxus/assets/25659432/bc53c084-5db1-46c5-8559-e36d2ff2679f">

calling types as functions `count()` - works internally because of this:
https://github.com/dtolnay/case-studies/blob/master/callable-types/README.md
https://github.com/DioxusLabs/dioxus/blob/ffe02b43c525123ff9ea79e9879d33517bd4059d/packages/signals/src/read.rs#L119
